### PR TITLE
[android] set presentation time via eglPresentationTimeANDROID

### PR DESCRIPTION
**Pull request by iskakaushik**
Attempt to reland: https://github.com/flutter/engine/pull/29727/

TODO:

1. This isn't right in the cases where the frame has been resubmitted.
2. This hasn't been wired for impeller yet.

### CODE REVIEW

**Comment by iskakaushik** [comment]
On a test application, this was seen to reduce the total time spent in `dequeueBuffer` from ~550ms to ~150ms.

1. [pprof prior to this change](https://pprof.corp.google.com/?id=c4fce2b47d78dfde3f4673c62fb3d80d&pivot=dequeueBuffer).
2. [pprof with this change](https://pprof.corp.google.com/?id=5f2e461e0a262a29e89a1c351bc42d92&path=1n3ky181mgb7o80ike6tk1eo7zmb1fjcfog1b7yltj0gh7f480p0nekg15mrzq81rgl6s0133jwoo1o86ax40os8pag066nxvk0jv04yc0d4r5zc1psb8uk09q4er40jfaang0ue4suo0lap90o0iw0xw81tpt27s0f2nx9w176tge41il4wtc1chdn240n85h6k0tf04go1ugrr600tiu36o00ny7ww1a39sgo0xy2das&pivot=dequeueBuffer)


cc: @dnfield

File: shell/common/rasterizer.cc [hunk 1]
```diff
@@ -604,6 +604,9 @@ RasterStatus Rasterizer::DrawToSurfaceUnsafe(
     }
 
     SurfaceFrame::SubmitInfo submit_info;
+    // TODO (kaushikiska): this can be in the past and might need to get snapped
+    // to future as this frame could have been resubmitted.
+    submit_info.presentation_time = frame_timings_recorder.GetVsyncTargetTime();
```
Comment (dnfield) [inline shell/common/rasterizer.cc:609]: Bug?

Should we avoid setting this if it's a resubmit?

File: shell/gpu/gpu_surface_gl_impeller.cc [hunk 2]
```diff
@@ -63,6 +63,8 @@ std::unique_ptr<SurfaceFrame> GPUSurfaceGLImpeller::AcquireFrame(
       GLPresentInfo present_info = {
           .fbo_id = 0,
           .damage = std::nullopt,
+          // TODO (kaushikiska): wire-up presentation time to impeller backend.
```
Comment (dnfield) [inline shell/gpu/gpu_surface_gl_impeller.cc:66]: Bug?

File: shell/platform/android/android_egl_surface.cc [hunk 3]
```diff
@@ -60,6 +60,17 @@ void LogLastEGLError() {
   FML_LOG(ERROR) << "Unknown EGL Error";
 }
 
+namespace {
+
+static bool HasExtension(const char* extensions, const char* name) {
```
Comment (dnfield) [inline shell/platform/android/android_egl_surface.cc:65]: We have this method in like 3 or 4 places now.  Or maybe we will with the swiftshader patch. We should consider moving it somewhere common

**Review by dnfield — COMMENTED** [review]
It'd be great to have some kind of test for this, especially if we have to special case resubmit frames.

File: shell/gpu/gpu_surface_gl_impeller.cc (hunk 2 above)
Comment (iskakaushik) [reply shell/gpu/gpu_surface_gl_impeller.cc]: done!

File: shell/common/rasterizer.cc (hunk 1 above)
Comment (iskakaushik) [reply shell/common/rasterizer.cc]: filed an issue. for now i'll set this only if it is `> now()`, I think it should be the case for majority of frames. Once the issue is addressed we can snap this to the next vsync target time,.

**Comment by iskakaushik** [comment]
> It'd be great to have some kind of test for this, especially if we have to special case resubmit frames.

Done! Added tests for cases where the frame is submitted in the future vs in past.

File: shell/platform/android/android_egl_surface.cc (hunk 3 above)
Comment (dnfield) [reply shell/platform/android/android_egl_surface.cc:65]: SGTM

File: shell/common/rasterizer_unittests.cc [hunk 4]
```diff
@@ -830,4 +831,152 @@ TEST(RasterizerTest,
...
+  std::string test_name =
+      ::testing::UnitTest::GetInstance()->current_test_info()->name();
+  ThreadHost thread_host("io.flutter.test." + test_name + ".",
+                         ThreadHost::Type::Platform | ThreadHost::Type::RASTER |
+                             ThreadHost::Type::IO | ThreadHost::Type::UI);
+  TaskRunners task_runners("test", thread_host.platform_thread->GetTaskRunner(),
+                           thread_host.raster_thread->GetTaskRunner(),
+                           thread_host.ui_thread->GetTaskRunner(),
+                           thread_host.io_thread->GetTaskRunner());
+  MockDelegate delegate;
```
Comment (dnfield) [inline shell/common/rasterizer_unittests.cc:844]: nit: use `NiceMock<MockDelegate>` (`using ::testing::NiceMock`) so this doesn't spam output about unused calls, here and elsewhere.

**Review by dnfield — APPROVED** [review]
LGTM with nit

File: shell/common/rasterizer_unittests.cc (hunk 4 above)
Comment (iskakaushik) [reply shell/common/rasterizer_unittests.cc:844]: nice! https://github.com/flutter/flutter/issues/105631, i think we want to use it everywhere in this file.

Comment (dnfield) [reply shell/common/rasterizer_unittests.cc:844]: I'm updating the rest of the file in https://github.com/flutter/engine/pull/33890 which is currently stuck on Fuchsia weirdness.

**Comment by zanderso** [comment]
Could this change have been responsible for the improvement to 99%-ile raster times in backdrop_filter_perf? https://flutter-flutter-perf.skia.org/e/?begin=1653514394&end=1659451447&keys=X29da4c759a0ed692aa4ac443fa2cfbb3&requestType=0&xbaroffset=30147

**Pull request merged**